Index: kernel-source/drivers/rtc/Kconfig
===================================================================
--- kernel-source.orig/drivers/rtc/Kconfig
+++ kernel-source/drivers/rtc/Kconfig
@@ -1202,6 +1202,12 @@ config RTC_DRV_ZYNQMP
 	  If you say yes here you get support for the RTC controller found on
 	  Xilinx Zynq Ultrascale+ MPSoC.
 
+config RTC_DRV_HA1588
+	tristate "HA1588 RTC IP Core"
+	depends on OF
+	help
+	  If you say yes here you get support for the HA1588 RTC Core
+
 comment "on-CPU RTC drivers"
 
 config RTC_DRV_ASM9260
Index: kernel-source/drivers/rtc/Makefile
===================================================================
--- kernel-source.orig/drivers/rtc/Makefile
+++ kernel-source/drivers/rtc/Makefile
@@ -68,6 +68,7 @@ obj-$(CONFIG_RTC_DRV_EP93XX)	+= rtc-ep93
 obj-$(CONFIG_RTC_DRV_FM3130)	+= rtc-fm3130.o
 obj-$(CONFIG_RTC_DRV_GEMINI)	+= rtc-gemini.o
 obj-$(CONFIG_RTC_DRV_GENERIC)	+= rtc-generic.o
+obj-$(CONFIG_RTC_DRV_HA1588)	+= rtc-ha1588.o
 obj-$(CONFIG_RTC_DRV_HID_SENSOR_TIME) += rtc-hid-sensor-time.o
 obj-$(CONFIG_RTC_DRV_HYM8563)	+= rtc-hym8563.o
 obj-$(CONFIG_RTC_DRV_IMXDI)	+= rtc-imxdi.o
